### PR TITLE
Model-owned inference context and versioned conversation-state snapshots

### PR DESCRIPTION
**Pull request by Avicennasis**
## What this does

Two changes toward a model-backend boundary, in dependency order. Stacked on #24 (on #23, #22, #21); the diff includes those until they merge — the four commits from `8ef1753` on are this PR.

### 1. Model-owned inference context (source-breaking)

`InferenceModel` no longer exposes Qwen-typed decode state. Models hand out an opaque `InferenceContext` they own and advance; the KV cache, conv history, DeltaNet recurrence, position, and the GPU-resident KV from #23 are internals of `QwenInferenceContext`. `QwenCPUModel.DecodeState` is off the public surface.

```swift
protocol InferenceContext: AnyObject { var position: Int { get }; func reset() }
protocol InferenceModel {
    var vocabSize: Int { get }; var modelDir: URL { get }
    func makeContext() -> any InferenceContext
    func step(_ ids: [Int], context: any InferenceContext) throws -> [Float]
}
```

A context is owned by the instance that created it; a foreign one (other instance, or a CPU context on a Metal model) throws `InferenceContextError.foreignContext`. Call-site churn (CLI, server, session, tests) is mechanical.

This also fixed a latent bug: on the Metal fast path the DeltaNet recurrence lives in model-owned GPU buffers bound to one state, so before this a *second* context used mid-sequence on the same model was silently wrong. Binding a context now captures the outgoing recurrence from the GPU and loads the incoming one's (host-side memcpy only, no dispatches). The interleaved-contexts test was red before that commit and is bitwise-green after.

Tests: a Qwen-free stub backend drives `TextGenerator` end to end (compile-level proof the protocol carries no Qwen types); the protocol file is source-scanned; foreign contexts are refused on both engines; create → step → reset → step is bitwise equal to a fresh context on CPU and Metal; interleaved contexts on one model are bitwise equal to isolated runs. All dispatch/command-buffer baselines and parity pins unchanged. `QwenMetalModel.swift` footprint: +11/−13 lines.

### 2. Conversation-state snapshots

`PersistableInferenceModel` + a versioned binary format (`SWLSTATE` v1, 128-byte header: magic, version, dtype, identity scheme, SHA-256 model identity, 11-value geometry table, position, section table, total length; sections for K, V, conv tail, delta state; trailing SHA-256 over all preceding bytes). Identity scheme 1 pins every weight byte via a qpack's `hashes.json`; scheme 2 (plain safetensors) pins structure only and says so. Strict refusals, each tested: bad magic, unsupported version, truncated, corrupt digest/section table, dtype mismatch, model mismatch, forged geometry, section-vs-position, foreign-context snapshot.

Correctness: save after N tokens → restore into a **fresh model instance** → continue: logits and greedy picks bitwise equal to the uninterrupted run on CPU and Metal, and across a layer-major chunked prefill. GPU KV round-trips through the CPU mirror and is proven equal to a direct GPU readback; conv-tail/delta sections likewise; snapshotting leaves the live context undisturbed. Mutation check: zeroing the restored delta state fails all four round-trip tests. Real-model spot check (35B qpack, base M1): 16 tokens + save (67 MB) → new process → 15 more tokens, id-for-id equal to the straight 32-token run.

CLI: `--save-state` / `--load-state`. Format documented in `docs/STATE_FORMAT.md`. Not persisted (documented): session-level message matching in `SwiftletSession`.

116 tests / 20 suites on a base M1 (CLT Swift 6.3.3). Rebase note vs #25: the only textual overlap is the `step(...)` signature line in `QwenMetalModel.swift`.

### CODE REVIEW

**Comment by leonickson1** [comment]
Merged the rest of the batch, thank you for all of it. Holding just this one for now, and I want to explain why so it does not read as arbitrary.

This is the source breaking one. Moving InferenceModel to a model owned InferenceContext and taking DecodeState off the public surface is the right direction, and I actually like the design. The opaque context and the versioned snapshot format are clean, and the latent second context bug you caught on the Metal fast path is a good find. The catch is that anything calling the model directly has to move to the new step signature, and I have an app sitting on top of this that I need to rebuild and retest against the new API in the same pass. I would rather do that deliberately than land the break in the middle of everything else going in.

To be clear this is not a quality concern. I read it and built it and it is green on my end. I will pick it up in the next round and most likely fold the app update into the same change so nothing is left half migrated. The conversation state snapshots in particular are something I want.

Leaving it open so it stays on my list. Thanks again, this whole set was a lot of careful work and the writeups made it easy to follow.
